File: packages/duet-js/src/typed/reading.ts
/**
 * What a typed field currently holds — including the two answers that are not a
 * value.
 *
 * Mirrors `packages/duet/lib/src/typed/duet_reading.dart`.
 *
 * @module
 */

import { echoBounded } from '../errors.ts';
import type { DuetValue } from '../value.ts';
import type { DuetCodec } from './codec.ts';

/** The path holds a value the codec accepted. */
export interface DuetPresent<T extends {}> {
  readonly kind: 'present';
  /** The decoded value. */
  readonly value: T;
}

/**
 * The path exists and holds a value of kind `'null'` — Rust's `Option::None`.
 *
 * Only {@link DuetOptionalField} produces this. A {@link DuetField} whose schema
 * promises a `T` reports a {@link DuetMismatch} for a null, because a null is
 * not the `T` it was promised.
 */
export interface DuetNone {
  readonly kind: 'none';
}

/**
 * There is no node at the path at all.
 *
 * Distinct from {@link DuetNone}: nothing was ever written here, or an ancestor
 * of this path is a scalar, or a list index is out of range. The host says this
 * with a `null` where a value would go; {@link DuetClient.get} says it with
 * `null`.
 */
export interface DuetAbsent {
  readonly kind: 'absent';
}

/**
 * The path holds a value, and the codec refused it.
 *
 * The honest report of two guests disagreeing about one path's type. Carries
 * what was actually found so an application can log it, show it, or write the
 * right type back over it.
 */
export interface DuetMismatch {
  readonly kind: 'mismatch';
  /** The value the host actually holds at the path. */
  readonly found: DuetValue;
  /**
   * Why the codec refused it, already bounded by {@link MAX_ECHO_CHARS} wherever
   * it embeds host-supplied text.
   */
  readonly reason: string;
}

/**
 * One typed observation of one path.
 *
 * # Why a tagged result and not an exception
 *
 * A type mismatch is a **state the host can be in**, not a failure of the call
 * that found it. Another guest can write any value to any path — the two-guest
 * proof has a webview and a Flutter engine writing one store at the same time —
 * so a typed watcher will eventually meet a value its codec refuses. That is
 * data, and it arrives through a push, where there is no call stack to throw
 * into and where throwing would take out the transport's message handler.
 *
 * Given that `watch` cannot throw, `get` **must not** throw either. Two
 * mechanisms for one condition would mean an application that handles a mismatch
 * on its watcher still crashes the first time it reads the same path directly.
 * One tagged type used by `get`, by `watch` and by the router's own mirror is
 * the smaller API: a reader learns it once, and a `switch` over `kind` narrows
 * exhaustively.
 *
 * # Why four arms
 *
 * Because the host can be in exactly four states, and the wire already spends a
 * distinct spelling on each: a decodable value, `{"t":"n"}`, a `null` where a
 * value would go, and a value of some other type. Collapsing any two would
 * delete a distinction the protocol pays for — the same argument
 * {@link decodeOptionalValue} makes one layer down.
 *
 * # Where this diverges from the Dart mirror, and why
 *
 * Dart's version is a `sealed` class hierarchy, so every arm must carry `T` for
 * the hierarchy itself to be `DuetReading<T>`. TypeScript's unions are
 * structural, so only the arm that actually holds a `T` is generic. The arm set,
 * the tags and the semantics are identical.
 */
export type DuetReading<T extends {}> = DuetPresent<T> | DuetNone | DuetAbsent | DuetMismatch;

/** Builds a present reading. */
export function duetPresent<T extends {}>(value: T): DuetPresent<T> {
  return { kind: 'present', value };
}

/** The one "explicitly null" reading. */
export function duetNone(): DuetNone {
  return { kind: 'none' };
}

/** The one "no such path" reading. */
export function duetAbsent(): DuetAbsent {
  return { kind: 'absent' };
}

/** Builds a mismatch reading. */
export function duetMismatch(found: DuetValue, reason: string): DuetMismatch {
  return { kind: 'mismatch', found, reason };
}

/**
 * The value, if `reading` is present; `null` otherwise.
 *
 * A convenience for the common "render it or render nothing" case. Prefer a
 * `switch` over `kind` where the three non-value outcomes differ.
 */
export function duetReadingValue<T extends {}>(reading: DuetReading<T>): T | null {
  return reading.kind === 'present' ? reading.value : null;
}

/**
 * A live typed subscription, returned by {@link DuetField.watch}.
 *
 * `current` is up to date the instant `watch` resolves: it already reflects the
 * host's subscription snapshot and every notification that raced ahead of the
 * reply. The callback fires only for changes *after* that point, so an
 * application never receives a notification for a handle it does not yet hold.
 */
export interface DuetWatch<T extends {}> {
  /** The most recent reading, updated before the callback for it runs. */
  readonly current: DuetReading<T>;

  /** True once {@link DuetWatch.close} has run. */
  readonly isClosed: boolean;

  /**
   * Cancels the subscription on the host and stops the callback.
   *
   * Idempotent. Rejects with whatever {@link DuetClient.unsubscribe} rejects
   * with, so a host that refuses the cancellation is not silently ignored.
   */
  close(): Promise<void>;
}

/**
 * Reads `value` as a required `T`.
 *
 * - `null` — no node at the path — is {@link DuetAbsent}.
 * - Anything the codec accepts is {@link DuetPresent}.
 * - Everything else, **including a `'null'` value**, is {@link DuetMismatch}: a
 *   required field promised a `T`, and a null is not one.
 */
export function duetRequiredReading<T extends {}>(
  codec: DuetCodec<T>,
  value: DuetValue | null,
): DuetReading<T> {
  if (value === null) return duetAbsent();
  return decodeOrMismatch(codec, value);
}

/**
 * Reads `value` as an optional `T`, mirroring Rust's `Option<T>`.
 *
 * - `null` — no node at the path — is {@link DuetAbsent}.
 * - A `'null'` value is {@link DuetNone}, without consulting the codec:
 *   `Option::None` lowers to `Value::Null` by definition, and no codec for a
 *   non-nullable `T` may claim it.
 * - Anything the codec accepts is {@link DuetPresent}; everything else is
 *   {@link DuetMismatch}.
 */
export function duetOptionalReading<T extends {}>(
  codec: DuetCodec<T>,
  value: DuetValue | null,
): DuetReading<T> {
  if (value === null) return duetAbsent();
  if (value.kind === 'null') return duetNone();
  return decodeOrMismatch(codec, value);
}

/**
 * Runs `codec` over `value`, turning every outcome into a reading.
 *
 * The `catch` is what makes this package's totality claim survive contact with
 * code it does not own. A codec is a **decoder**, and every decode path in this
 * package is total; a generated or hand-written codec that throws would
 * otherwise put an exception on the push path, where {@link DuetClient} hands it
 * straight to the transport's message handler.
 *
 * This is deliberately the *opposite* of {@link DuetClient.onPush}'s policy,
 * which lets an application handler's exception escape. The difference is whose
 * bug it is and whether it can be reported: a throwing codec is reported here,
 * as a mismatch carrying the thrown text, so it is visible rather than swallowed
 * — an application callback has no such channel, and hiding its exception would
 * hide a bug in the one place nobody would look.
 */
function decodeOrMismatch<T extends {}>(
  codec: DuetCodec<T>,
  value: DuetValue,
): DuetReading<T> {
  let decoded: T | null;
  try {
    decoded = codec.decode(value);
  } catch (error) {
    return duetMismatch(
      value,
      `the ${codec.name} codec threw: ${echoBounded(String(error))}`,
    );
  }
  if (decoded === null) {
    return duetMismatch(
      value,
      `expected ${codec.name}, found a value of kind ${echoBounded(value.kind)}`,
    );
  }
  return duetPresent(decoded);
}

File: crates/duet-schema/src/typed/reading.rs
//! What a typed field currently holds — including the answers that are not a
//! value.

use duet_core::Value;

use crate::decode::DecodeError;
use crate::state::{NotNullable, SharedState};

/// One typed observation of one path.
///
/// # Why a result type and not an error
///
/// A type mismatch is a **state the host can be in**, not a failure of the call
/// that found it. Another guest can write any value to any path — the two-guest
/// proof has a webview and a Flutter engine writing one store at the same time
/// — so a typed read will eventually meet a value its type refuses. That is
/// data. `Field::get` still returns a `Result`, but its `Err` is reserved for
/// the store being unreachable, which is a genuinely different condition and
/// one the application cannot inspect its way out of.
///
/// # Why four arms
///
/// Because the host can be in exactly four states and the wire already spends a
/// distinct spelling on each: a decodable value, `Value::Null`, no node at all,
/// and a value of some other type. Collapsing any two would delete a
/// distinction the protocol pays for.
///
/// These are the same four the Dart and TypeScript clients settled on
/// (`DuetPresent` / `DuetNone` / `DuetAbsent` / `DuetMismatch`), deliberately:
/// the whole point of the schema is that all three languages agree about what a
/// path holds, and they cannot agree if they disagree about what "empty" means.
#[derive(Debug, Clone, PartialEq)]
pub enum Reading<T> {
    /// The path holds a value the type accepted.
    Present(T),
    /// The path exists and holds [`Value::Null`] — Rust's `Option::None`.
    ///
    /// Only an [`OptionalField`](crate::OptionalField) produces this. A
    /// [`Field`](crate::Field) whose schema promises a `T` reports
    /// [`Reading::Mismatch`] for a null, because a null is not the `T` it was
    /// promised.
    None,
    /// There is no node at the path at all.
    ///
    /// Distinct from [`Reading::None`]: nothing was ever written here, or an
    /// ancestor of this path is itself `Null` or a scalar, or a list index is
    /// out of range.
    Absent,
    /// The path holds a value, and the type refused it.
    ///
    /// The honest report of two guests disagreeing about one path. Carries what
    /// was actually found, so an application can log it, show it, or write the
    /// right type back over it.
    Mismatch {
        /// The value the store actually holds at the path.
        found: Value,
        /// Why it was refused.
        error: DecodeError,
    },
}

impl<T> Reading<T> {
    /// The value if this is [`Reading::Present`], `None` otherwise.
    ///
    /// A convenience for the common "render it or render nothing" case. Prefer
    /// an exhaustive `match` wherever the three non-value outcomes differ —
    /// treating a mismatch as an absence is how a type disagreement between two
    /// guests becomes invisible.
    pub fn present(self) -> Option<T> {
        match self {
            Reading::Present(value) => Some(value),
            _ => None,
        }
    }

    /// True for [`Reading::Present`].
    pub fn is_present(&self) -> bool {
        matches!(self, Reading::Present(_))
    }

    /// True for [`Reading::Mismatch`].
    pub fn is_mismatch(&self) -> bool {
        matches!(self, Reading::Mismatch { .. })
    }
}

/// Reads `value` as a required `T`.
///
/// - `None` — no node at the path — is [`Reading::Absent`].
/// - Anything `T` accepts is [`Reading::Present`].
/// - Everything else, **including [`Value::Null`]**, is [`Reading::Mismatch`]:
///   a required field promised a `T`, and a null is not one.
pub fn required_reading<T: SharedState>(value: Option<Value>) -> Reading<T> {
    let Some(found) = value else {
        return Reading::Absent;
    };
    decode_or_mismatch(found)
}

/// Reads `value` as an optional `T`, mirroring Rust's `Option<T>`.
///
/// - `None` — no node at the path — is [`Reading::Absent`].
/// - [`Value::Null`] is [`Reading::None`], **without consulting `T`**:
///   `Option::None` lowers to `Value::Null` by definition, and the
///   [`NotNullable`] bound guarantees no `T` here may claim it.
/// - Anything else `T` accepts is [`Reading::Present`]; the rest is a mismatch.
pub fn optional_reading<T: SharedState + NotNullable>(value: Option<Value>) -> Reading<T> {
    match value {
        None => Reading::Absent,
        Some(Value::Null) => Reading::None,
        Some(found) => decode_or_mismatch(found),
    }
}

/// Runs `T`'s decoder over `found`, turning either outcome into a reading.
///
/// No `catch` equivalent, and none is needed: unlike the Dart and TypeScript
/// clients — which must tolerate an application-supplied codec that throws —
/// `SharedState::from_value` returns a `Result` and this crate's own impls are
/// total by construction. A hand-written impl that panics takes the process
/// down, which is the correct outcome for a bug in a decoder rather than
/// something to paper over.
fn decode_or_mismatch<T: SharedState>(found: Value) -> Reading<T> {
    match T::from_value(&found) {
        Ok(value) => Reading::Present(value),
        Err(error) => Reading::Mismatch { found, error },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_missing_node_is_absent_for_both_kinds_of_field() {
        assert_eq!(required_reading::<i64>(None), Reading::Absent);
        assert_eq!(optional_reading::<i64>(None), Reading::Absent);
    }

    #[test]
    fn a_decodable_value_is_present() {
        assert_eq!(
            required_reading::<i64>(Some(Value::Int(7))),
            Reading::Present(7)
        );
        assert_eq!(
            optional_reading::<i64>(Some(Value::Int(7))),
            Reading::Present(7)
        );
    }

    #[test]
    fn null_splits_the_two_readings() {
        // The distinction the whole typed layer exists to keep: a required
        // field promised a value, so a null is a disagreement; an optional
        // field promised `Option`, so a null is the answer.
        assert!(required_reading::<i64>(Some(Value::Null)).is_mismatch());
        assert_eq!(optional_reading::<i64>(Some(Value::Null)), Reading::None);
    }

    #[test]
    fn a_wrong_typed_value_is_a_mismatch_carrying_what_was_found() {
        let reading = required_reading::<i64>(Some(Value::Str("nope".into())));
        match reading {
            Reading::Mismatch { found, error } => {
                assert_eq!(found, Value::Str("nope".into()));
                assert_eq!(error.to_string(), "expected i64 at <root>, found string");
            }
            other => panic!("expected a mismatch, got {other:?}"),
        }
    }

    #[test]
    fn absent_and_none_are_not_equal() {
        // Pinned because collapsing them is the single most tempting
        // simplification here, and it would silently delete the distinction
        // the wire pays a spelling for.
        assert_ne!(Reading::<i64>::Absent, Reading::None);
    }

    #[test]
    fn present_extracts_only_the_value_arm() {
        assert_eq!(Reading::Present(7i64).present(), Some(7));
        assert_eq!(Reading::<i64>::None.present(), None);
        assert_eq!(Reading::<i64>::Absent.present(), None);
        assert_eq!(
            Reading::<i64>::Mismatch {
                found: Value::Null,
                error: DecodeError::wrong_type("i64", &Value::Null),
            }
            .present(),
            None
        );
    }

    #[test]
    fn the_predicates_agree_with_the_arms() {
        assert!(Reading::Present(1i64).is_present());
        assert!(!Reading::<i64>::Absent.is_present());
        assert!(!Reading::<i64>::None.is_present());
        assert!(!Reading::Present(1i64).is_mismatch());
        assert!(
            Reading::<i64>::Mismatch {
                found: Value::Null,
                error: DecodeError::wrong_type("i64", &Value::Null),
            }
            .is_mismatch()
        );
    }

    #[test]
    fn a_struct_shaped_reading_survives_every_hostile_variant() {
        for value in [
            Value::Null,
            Value::Bool(true),
            Value::Int(1),
            Value::Float(1.0),
            Value::Str("x".into()),
            Value::Bytes(vec![1]),
            Value::List(vec![Value::Null]),
            Value::map([("a", Value::Null)]),
        ] {
            let _ = required_reading::<Vec<i64>>(Some(value.clone()));
            let _ = optional_reading::<Vec<i64>>(Some(value));
        }
    }
}
